### PR TITLE
[latest] publish built-ins to Open VSX registry

### PR DESCRIPTION
**Pull request by marcdumais-work**
Added /src/publish-vsix.js to permit publishing to an Open VSX registry.
First run yarn to build the builtin extensions as .vsix in folder 'dist'
and remove any not wished published (e.g. not working well).

Then run:
   $> yarn publish:latest

To publish to a specific registry, use environment variable
OVSX_REGISTRY_URL to point to it. Defaults to open-vsx.org

Use environment variable OVSX_PAT to store your access token for the
registry.

Also improved the auto-generated README used in our .vsix, so it's clearer
that Microsoft is not involved in any way other than being the author of
these extensions we package and publish.

Added a copy of the VS Code license file for each builtin.

For vsix packaging, now use version number that reflects the VS Code
version as per its package.json.

Signed-off-by: Marc Dumais <marc.dumais@ericsson.com>

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -6,7 +6,8 @@
         "build:extensions": "yarn --cwd vscode && yarn compile:extensions",
         "compile:extensions": "NODE_OPTIONS=--max-old-space-size=8192 node ./src/compile.js",
         "bundle:extensions": "NODE_OPTIONS=--max-old-space-size=8192 node ./src/bundle.js",
-        "publish:latest": "node ./src/publish.js --tag=latest",
+        "publish": "yarn && yarn publish:latest",
+        "publish:latest": "node ./src/publish-vsix.js --tag=latest",
         "publish:next": "node ./src/publish.js --tag=next",
```
Comment (marcdumais-work) [inline package.json:11]: "next" left untouched for now - we have no way yet to publish a next on open-vsx AFAIK

File: src/package-vsix.js [hunk 2]
```diff
@@ -17,26 +17,26 @@ const repository = {
     "url": "https://github.com/theia-ide/vscode-builtin-extensions"
 };
 
-// bump to publish
-let version = '1.39.1-prel';
+const vscodePck = JSON.parse(fs.readFileSync(vscode('package.json'), 'utf-8'));
+let version = vscodePck.version || '0.0.1';
```
Comment (marcdumais-work) [inline src/package-vsix.js:21]: when vsix packaging a `latest`, the version of the builtins is taken from vscode's package.json

File: src/package-vsix.js [hunk 3]
```diff
@@ -17,26 +17,27 @@ const repository = {
...
     const bin = await run('yarn', ['bin'], root());
     const vsce = bin.trim() + '/vsce';
 
+    // use VS Code version and SHA when publishing next
     if (tag === 'next') {
         const shortRevision = (await run('git', ['rev-parse', '--short', 'HEAD'], vscode())).trim();
-        const [, minor] = version.split('.');
-        version = `0.${Number(minor) + 1}.0-next.${shortRevision}`;
+        const [major, minor , ] = version.split('.');
+        version = `${major}.${Number(minor) + 1}.0-next.${shortRevision}`;
```
Comment (marcdumais-work) [inline src/package-vsix.js:31]: @spoenemann do you think this way of versioning a `next` or `insider` will work ok when publishing on open-vsx.org? e.g. for the current vscode `1.44.2` it will give a version like `1.45.0-next.ff91584411`

File: src/publish-vsix.js [hunk 4]
```diff
@@ -0,0 +1,42 @@
...
+const yargs = require('yargs');
+const { root, dist, run } = require('./paths.js');
+
+const { tag } = yargs.option('tag', {
+    choices: ['latest', 'next']
+}).demandOption('tag').argv;
+
+(async () => {
+    if (tag === 'next') {
+        console.error("Open VSX does not support publishing 'next' versions at this time");
```
Comment (marcdumais-work) [inline src/publish-vsix.js:25]: @spoenemann  I do not think we have a way yet to have `latest` and `next` versions co-habitate on open-svx and being selectively consumable in a Theia product? 

I did a little research and though requested for a long time this is not supported on the Marketplace, either. So MS extensions that have `insiders` versions do not have a good mechanism to do that. For example `ms-vscode.cpptools` has its own update mechanism that pulls an `insiders` version of itself from their GH release, and subsequent updates I think too, if a preference to that effect is enabled.

**Comment by marcdumais-work** [comment]
FYI, @spoenemann , @akosyakov . This is what I used to build and publish builtins `1.44.2` earlier today.

File: src/package-vsix.js (hunk 3 above)
Comment (spoenemann) [reply src/package-vsix.js:31]: A `next` version should also set `preview` to `true` in its package.json, but then I think it's fine.

Currently the Open VSX website displays the latest published version, including preview versions. We need to think about whether we want to go in the direction of npmjs.com: display the `latest` version (excluding next/preview) as default, but allow to choose any other version in the UI.

File: src/publish-vsix.js (hunk 4 above)
Comment (spoenemann) [reply src/publish-vsix.js:25]: Yes the MS Marketplace has no concept of version tags, only a `preview` flag. See my comment above: going for the npm way would give us more flexibility, but it also means the UI would become more complex.

File: src/package-vsix.js [hunk 5]
```diff
@@ -78,24 +79,25 @@ let version = '1.39.1-prel';
...
         // lookup extensions, and so changing it may prevent dependent extensions
-        // to not find it
-        // pck.displayName = nls.displayName || extDisplayName + " (built-in)";
-        pck.displayName = nls.displayName? nls.displayName + " (built-in)" : extDisplayName ;
+        // from finding it
+        pck.displayName = nls.displayName ? nls.displayName + " (built-in)" : extDisplayName;
         pck.description = nls.description || "Built-in extension that adds (potentially basic) support for " + capitalize(pck.name);
         pck.keywords = ["Built-in"];
         pck.repository = repository;
         pck.version = version;
```
Comment (spoenemann) [inline src/package-vsix.js:90]: We should set
```
pck.license = 'SEE LICENSE IN LICENSE-vscode.txt';
```
so the registry knows where to find the license file. An alternative would be to rename the file to the standard name `LICENSE`.

File: src/package-vsix.js [hunk 6]
```diff
@@ -106,19 +108,32 @@ let version = '1.39.1-prel';
...
+## Disclaimer
     
+Microsoft does not endorse, build, test or publish this extension, nor are they involved with it in any other way. The only association is that they own the copyright of these extensions and the rest of vscode's [source code](https://github.com/microsoft/vscode/tree/master/extensions), which they released under the [MIT License](https://github.com/microsoft/vscode/blob/master/LICENSE.txt). A copy of the license is included in this extension. See LICENSE-vscode.txt
+
+    "Original VS Code sources are Copyright (c) 2015 - present Microsoft Corporation."
+
+
+## What is this extension? Do I need it?
+
+TL;DR: If you are running \`VS Code\`, \`Code OSS\` or derived product built from the VS Code repository, such as [VSCodium](https://github.com/VSCodium/vscodium), you do not need to install this extension since it's already present - "builtin".
```
Comment (spoenemann) [inline src/package-vsix.js:131]:
```suggestion
TL;DR: If you are running \`VS Code\`, \`Code OSS\` or a derived product built from the VS Code repository, such as [VSCodium](https://github.com/VSCodium/vscodium), you do not need to install this extension since it's already present - "built-in".
```

File: src/package-vsix.js [hunk 7]
```diff
@@ -106,19 +108,32 @@ let version = '1.39.1-prel';
...
+
+Built-in extensions are built-along and included in \`VS Code\` and \`Code OSS\`. In consequence they may be expected to be present and used by other extensions. They are part of the [vscode GitHub repository](https://github.com/microsoft/vscode/tree/master/) and generally contribute basic functionality such as textmate grammars, used for syntax-highlighting, for some of the most popular programming languages. In some cases, more substantial features are contributed through built-in extensions (e.g. Typescript, Markdown, git, ...). Please see the description above to learn what this specific extension does.
+
+To learn more about builtin extensions, including how they are built and packaged, please see [vscode-builtin-extensions](https://github.com/theia-ide/vscode-builtin-extensions)
```
Comment (spoenemann) [inline src/package-vsix.js:135]:
```suggestion
To learn more about built-in extensions, including how they are built and packaged, please see [vscode-builtin-extensions](https://github.com/theia-ide/vscode-builtin-extensions).
```

File: src/package-vsix.js (hunk 5 above)
Comment (marcdumais-work) [reply src/package-vsix.js:90]: done

File: src/package-vsix.js (hunk 6 above)
Comment (marcdumais-work) [reply src/package-vsix.js:131]: done

File: src/package-vsix.js (hunk 7 above)
Comment (marcdumais-work) [reply src/package-vsix.js:135]: done

File: src/package-vsix.js (hunk 3 above)
Comment (marcdumais-work) [reply src/package-vsix.js:31]: > We need to think about whether we want to go in the direction of npmjs.com: display the latest version (excluding next/preview) as default, but allow to choose any other version in the UI.

Yes, that's my main worry - whether we should do that and how. 

For now I have added the `preview=true` flag when packaging "next" versions of builtins, and I have made "next" the default when one calls "yarn".

File: package.json (hunk 1 above)
Comment (marcdumais-work) [reply package.json:11]: Correction as per discussion with Miro below: it is actually possible to publish next versions on open-vsx, it's just not yet possible to distinguish between "next" and "latest" - they are done the same way and "publish:latest" above can be used to publish either a "latest" or "next" to open-vsx. So conceptually this is closer to "publish:openvsx" and "publish:npm". 

In any case we can keep "publish:next" above until we are sure we no longer need to publish to npm.

**Comment by marcdumais-work** [comment]
> Great!
> 
> Could we put an Open VSX access token of the @open-vsx user into a GH secret here so we publish the built-ins with the same user as https://github.com/open-vsx/publish-extensions ?

Sure - can you try? I gave you write access to the repo.

**Comment by marcdumais-work** [comment]
@akosyakov are you ok too with merging this PR?
